test/missing.jl

```diff
@@ -0,0 +1,225 @@
+    for f in rounding_functions
+        @test ismissing(f(missing))
+        @test ismissing(f(missing, 1))
+        @test ismissing(f(missing, 1, 1))
+        @test ismissing(f(Union{Int, Missing}, missing))
+        @test_throws MissingException f(Int, missing)
+    end
+end
+
+@testset "printing" begin
+    @test sprint(show, missing) == "missing"
+    @test sprint(showcompact, missing) == "missing"
+    @test sprint(show, [missing]) == "$Missing[missing]"
+    @test sprint(show, [1 missing]) == "$(Union{Int, Missing})[1 missing]"
+    b = IOBuffer()
+    display(TextDisplay(b), [missing])
+    @test String(take!(b)) == "1-element Array{$Missing,1}:\n missing"
+    b = IOBuffer()
+    display(TextDisplay(b), [1 missing])
+    @test String(take!(b)) == "1×2 Array{$(Union{Int, Missing}),2}:\n 1  missing"
+end
+
+@testset "arrays with missing values" begin
+    x = convert(Vector{Union{Int, Missing}}, [1.0, missing])
+    @test isa(x, Vector{Union{Int, Missing}})
+    @test isequal(x, [1, missing])
+    x = convert(Vector{Union{Int, Missing}}, [1.0])
+    @test isa(x, Vector{Union{Int, Missing}})
+    @test x == [1]
+    x = convert(Vector{Union{Int, Missing}}, [missing])
+    @test isa(x, Vector{Union{Int, Missing}})
+    @test isequal(x, [missing])
+end
+
+@testset "== and != on arrays" begin
+    @test ismissing([1, missing] == [1, missing])
+    @test ismissing(["a", missing] == ["a", missing])
+    @test ismissing(Any[1, missing] == Any[1, missing])
+    @test ismissing(Any[missing] == Any[missing])
+    @test ismissing([missing] == [missing])
+    @test ismissing(Any[missing, 2] == Any[1, missing])
+    @test ismissing([missing, false] == BitArray([true, false]))
+    @test ismissing(Any[missing, false] == BitArray([true, false]))
+    @test Union{Int, Missing}[1] == Union{Float64, Missing}[1.0]
+    @test Union{Int, Missing}[1] == [1.0]
+    @test Union{Bool, Missing}[true] == BitArray([true])
+    @test !(Union{Int, Missing}[1] == [2])
+    @test !([1] == Union{Int, Missing}[2])
+    @test !(Union{Int, Missing}[1] == Union{Int, Missing}[2])
+
+    @test ismissing([1, missing] != [1, missing])
+    @test ismissing(["a", missing] != ["a", missing])
+    @test ismissing(Any[1, missing] != Any[1, missing])
+    @test ismissing(Any[missing] != Any[missing])
+    @test ismissing([missing] != [missing])
+    @test ismissing(Any[missing, 2] != Any[1, missing])
+    @test ismissing([missing, false] != BitArray([true, false]))
+    @test ismissing(Any[missing, false] != BitArray([true, false]))
+    @test !(Union{Int, Missing}[1] != Union{Float64, Missing}[1.0])
+    @test !(Union{Int, Missing}[1] != [1.0])
+    @test !(Union{Bool, Missing}[true] != BitArray([true]))
+    @test Union{Int, Missing}[1] != [2]
+    @test [1] != Union{Int, Missing}[2]
+    @test Union{Int, Missing}[1] != Union{Int, Missing}[2]
+end
+
+@testset "any & all" begin
+    @test any([true, missing])
```

Comment:
Should the result of `any(x)`/`all(x)` be still inferred as `Bool` when `x` cannot contain missing values? (IMO it would be a nice thing to have)
If so, maybe it could be tested here as well?

Reply:
Good idea. I've added `@inferred` everywhere it makes sense to test/reduce.jl rather than to this file, since it's more logical to test the behavior where type stability is expected. Also, benchmarks should should be pretty good at catching this kind of issue.